packages/http/src/mocker/negotiator/__tests__/NegotiatorHelpers.spec.ts

```diff
@@ -536,11 +536,14 @@ describe('NegotiatorHelpers', () => {
         headers: [],
       };
 
-      expect(() => {
-        helpers.negotiateDefaultMediaType(partialOptions, response);
-      }).toThrow(
-        'Could not generate response for provided content type or no content type provided. Tried to fallback to application/json, but no definition found.'
-      );
+      expect(helpers.negotiateDefaultMediaType(partialOptions, response)).toEqual({
```

Comment:
Please read my other comment first about this change and if we decide to push this forward please add an integration test.

For all future commits I'm thinking about adding this as a mandatory step. WDYT @StefanDywersant @philsturgeon @XVincentX ? We have missed too many bugs because we didn't have them.

Reply:
I'd be more cautious with integration tests. They are slow so we should do integration test for every tiny feature. Or we should establish a framework which lets use decide when to run which tests (units on every occasion, integration not).

We don't want to end with a tests running 10mins on local machine (@chris-miaskowski remember callinize server?)

Reply:
@StefanDywersant we can cherry pick integration tests. Right now we can't really complain on abundance of integration tests so I'd them to be added.

Reply:
So, from what you are saying, now we need to add integration _and_ unit tests _always_, for every change ?

Reply:
And btw. I can't add full integration test here. The same reason as with SO-82. Can't prepare jest environment to run a webserver, do a request and exit cleanly.

Reply:
Right, instead of approaching the test from:

1. CLI starts a web server
2. web server figures out which HTTP response to return
3. we test the web server has the right response

We skip step 1 and test that the http library is doing the right thing. This is actually when you _do_ want mocks. The tests on the CLI only need to confirm that when we put in --foo=abc that it passes on the correct things to the http library, or errors as we expect to, etc. Trying to do all the things via the CLI tests wil become a mess.

Reply:
> And btw. I can't add full integration test here. The same reason as with SO-82. Can't prepare jest environment to run a webserver, do a request and exit cleanly.

We already have integration tests, we call them "functional" tests. It's find if you just add them on the `http-server` layer.

> So, from what you are saying, now we need to add integration and unit tests always, for every change ?

I think it would make sense until we have a solid base of integration tests.

Reply:
What does it mean "solid base of integration tests" ?

I have a feeling that if we add integration tests to every small thing we do we end with multiplication of every unit test case for every component used there. And we invest lots of time. Both together is bad ;)

Also remember that we have full coverage of those cases via unit tests. Full!

I'm opposing because I think we should have reasonable amount of tests for each kind (~100% units, <100% func, <<100% integration). In this particular case I don't see any added value of duplicating the unit test with func one. But your single next comment is enough for me to feel overvoted and just do it ;)

Reply:
This conversation is getting a bit off the rails. 

@chris-miaskowski Yep, we have tests called functional, e.g: packages/http/src/mocker/__tests__/functional.spec.ts

Good. These tests seem to be unit tests to me, we put a variety of things in and see what comes out. I think your definition of integration test is just something which doesnt have any mocks or stubs, so by that definition yes these are integration tests.

@StefanDywersant we definitely dont want to be repeating unit tests and integration tests for everything, correct. I think you are using a similar definition of integration tests to me, which is something which hits multiple layers in a more realistic way.

Ignore the term "integration test" for a minute, Chris is correct that we need tests to confirm the new functionality, but this should be done at the appropriate boundary. Ignoring the CLI and the HTTP server, we want to make sure the HTTP mocker is doing the right thing. 

Karol has said we cannot (for technical reasons) have tests which start at the CLI, launch a server, then see what the server is responding with under specific situations, which is fine, because that's mostly a job for unit tests. 

We do however need tests which show that things are being passed from boundary to boundary correctly. 

- Unit tests for CLI which show it is talking to the HTTP server correctly under various circumstances
- Unit tests which show HTTP server is talking to HTTP correctly under various circumstances

Then a handful of integration tests which show the CLI can spin up a working instance, quickly poke a few things over HTTP, and close it down again. I would call these integration tests, and we only need a small number of them to confirm the CLI is doing its job, because the unit tests are confirming everything else. 

Sound good?

Reply:
The problem is that we have 100% unit test coverage and Phil keeps finding bugs. There is even a story @philsturgeon created specifically to create integration tests.

We now have so little integration tests that before we reach 'solid base' there will be plenty of time to move away from this idea.

Reply:
Ahhh, we are agreeing with different words.

Yes our unit tests are mocking far too much which is why we keep getting errors. 

You are asking for more integration tests, I am asking for unit tests that mock way less, these two things are the same but we have slightly different definitions of the term integration test.

Reply:
Also 100% to more integration tests for things that are not technically awful. The "CLI spawns web server" is hard, but "less mocky unit tests or integration tests whatever you call them" for everything else is definitely something we need.

Reply:
I totally agree with adding more functionals and more integration tests.
But I also want them to be good quality. Not just adding "on every occasion". This time the test it a total duplicate. All integration-related-stuff is tested in many ways and we are just changing the spec (which is exactly covered by unit test).

But to keep the train running, I just added the "test of contention" ;)

Reply:
@philsturgeon yes, integration tests to me (and I think by the definition) are tests that do not mock but are still testing code. They may seem similar to unit tests (which are 100% mocked) but they aren't.

What you are talking about is an e2e test, also known as functional tests: where you spin up the whole product and test similarly to how a user would test.